modules/porous_flow/include/userobjects/PorousFlowCount.h

```diff
@@ -0,0 +1,43 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef POROUSFLOWCOUNTUO_H
+#define POROUSFLOWCOUNTUO_H
+
+// MOOSE includes
+#include "ElementUserObject.h"
+
+// Forward Declarations
+class PorousFlowCount;
+
+template <>
+InputParameters validParams<PorousFlowCount>();
+
+/**
+ * doco
+ */
+class PorousFlowCount : public ElementUserObject
+{
+public:
+  PorousFlowCount(const InputParameters & parameters);
+  ~PorousFlowCount();
+
+  virtual void initialize() override;
+  virtual void execute() override;
+  virtual void finalize() override;
+  virtual void threadJoin(const UserObject & uo) override;
+  unsigned maxUpDownSwaps(unsigned ph) const;
+  void setProtoFlux(const std::vector<std::vector<Real>> & proto_flux);
+  Real getProtoFlux(unsigned ph, unsigned i) const;
+
+protected:
+  std::vector<std::vector<std::vector<unsigned>>> _num_upwinds;
+  std::vector<std::vector<std::vector<unsigned>>> _num_downwinds;
+  std::vector<std::vector<std::vector<Real>>> _proto_flux;
```

Comment:
I think I'd just go ahead and store these data structures right in your kernel. If you don't need to aggregate that information (`threadJoin()` and `finalize()`) but I don't think that's a problem for your use.